refactor: [M3-7399] - Ensure EU consent box shows for new European countries

jaalah-akamai opened the pull request:
## Description 📝
If a user has not signed the EU Model Contract Agreement, Cloud Manager prompts the user to sign it when they attempt to deploy a service in a DC whose slug starts with "eu-". This logic does not hold up with our new DC slug scheme, which means some DCs are missing the agreement.

**With these changes, when a user selects any region in EU, a GDPR checkbox should appear.**

## Changes  🔄
-  Created several utility functions to extract the information we need, but `getGDPRDetails` will provide us with the boolean we need to determine whether to hide/show the checkbox.
- Fixed a minor bug in `RegionSelect.tsx` where the selected region was not clearing when changing the create type
    - Extrapolated some of the logic from `RegionSelect.utils.ts` to make it more re-useable.
- GDPR Checkboxes should not be checked by default - we want the user to make a conscious decision to agree.

## Preview 📷
| Placement  | Screenshot   |
| ------- | ------- |
| Migration | ![Screenshot 2023-11-14 at 9 58 06 AM](https://github.com/linode/manager/assets/125309814/acfb4fd5-446b-4f8d-bc03-b49a1e74a8c0) |
| Image Upload | ![Screenshot 2023-11-14 at 9 58 24 AM](https://github.com/linode/manager/assets/125309814/2cdb10d1-9d79-4014-b285-784112eea012) |
| Linode Create | ![Screenshot 2023-11-14 at 10 00 56 AM](https://github.com/linode/manager/assets/125309814/598b6623-f1e7-40a3-ba3f-3352ceaecea7) |
| Nodebalancer Create | ![Screenshot 2023-11-14 at 10 16 21 AM](https://github.com/linode/manager/assets/125309814/57426792-5103-4cc4-b7db-f801d1725512) |
| Object Storage | ![Screenshot 2023-11-14 at 10 41 02 AM](https://github.com/linode/manager/assets/125309814/ed73f8aa-772b-43ce-998a-3f68726f069e) |
| Volumes Create | ![Screenshot 2023-11-14 at 10 45 36 AM](https://github.com/linode/manager/assets/125309814/0a3412b7-1ad4-4bae-81aa-9186b4b8166d) |
| Kubernetes Create | ![Screenshot 2023-11-14 at 12 09 55 PM](https://github.com/linode/manager/assets/125309814/bee4b311-a035-42d3-809a-6f295b8f9825) |

## How to test 🧪
> [!note]
Our Akamai accounts default `eu_model` to true you'll need to test on a personal account not tied to Akamai.

### Prerequisites
- Pull down branch or preview link

### Reproduction steps
- Go to `https://cloud.linode.com/linodes/create` and observe that most EU countries are missing GDPR. This applies to many other places throughout the app.

### Verification steps 
- [ ] When selected any region selector, ensure that the GDPR checkbox appears for European countries
- [ ] `/linodes/create` (and all the tabs for different create types)
    - [ ] Region selector should clear it's selected state when changing pages
- [ ] When migrating a linode
- [ ] `/volumes/create`
- [ ] `/nodebalancers/create`
- [ ] `/images/create/upload`
- [ ] `/kubernetes/create`
- [ ] `/object-storage/buckets/create`

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [x] 👣 Providing comprehensive reproduction steps
- [x] 📑 Providing or updating our documentation
- [x] 🕛 Scheduling a pair reviewing session
- [x] 📱 Providing mobile support
- [x] ♿  Providing accessibility support

abailly-akamai commented:
@jaalah-akamai will dig deeper into this PR but so far I noticed that the CreateBucketDrawer is broken as a result /object-storage/buckets

bnussman-akamai commented on packages/manager/src/utilities/formatRegion.ts:
```diff
@@ -19,5 +26,65 @@ export const getRegionCountryGroup = (region: Region | undefined) => {
+
+/**
+ *
+ * @param opts.agreements The user's agreements
+ * @param opts.profile The user's profile
+ * @param opts.regions The list of regions
+ * @param opts.selectedRegionId The ID of the selected region (e.g. 'eu-west')
```
Instead of defining all of these `@param`s, we could move the corresponding comments to the `GDPRConfiguration` interface. I think it would improve the experience for our IDEs

bnussman-akamai commented on packages/manager/src/utilities/formatRegion.ts:
```diff
@@ -19,5 +26,65 @@ export const getRegionCountryGroup = (region: Region | undefined) => {
+ * @param opts.profile The user's profile
+ * @param opts.regions The list of regions
+ * @param opts.selectedRegionId The ID of the selected region (e.g. 'eu-west')
+ * @returns selectedRegionGroup: string | undefined - the group of the selected region, if it exists
```
I think it's best practice to omit the types (`string | undefined`) here and let the language server handle it

bnussman-akamai commented on packages/manager/src/utilities/formatRegion.ts:
```diff
@@ -19,5 +26,65 @@ export const getRegionCountryGroup = (region: Region | undefined) => {
+ * @param opts.selectedRegionId The ID of the selected region (e.g. 'eu-west')
+ * @returns selectedRegionGroup: string | undefined - the group of the selected region, if it exists
+ * @returns showGDPRCheckbox: boolean - whether or not to show the GDPR checkbox
+ * @usage
```
`@example` will allow VSCode to add syntax highlighting

```suggestion
 * @example
```

bnussman-akamai commented on packages/manager/src/utilities/formatRegion.ts:
```diff
@@ -4,7 +4,14 @@ import {
+import type { Agreements, Country, Profile } from '@linode/api-v4';
+
+interface GDPRConfiguration {
+  agreements?: Agreements;
+  profile?: Profile;
```
Should we switch to 
```suggestion
  agreements: Agreements | undefined;
  profile: Profile | undefined;
```

so that profile and agreements are required to be passed even though they may be undefined?

abailly-akamai commented:
@jaalah-akamai the node balancer create flow is broken as well /nodebalancers/create
Looks like a state issue - the create button won't get enabled at all unless selecting an EU region and agreeing to terms


<video src="https://github.com/linode/manager/assets/130582365/6a732d9e-5644-461b-b10f-8ba8f56090fd" />

dwiley-akamai commented on packages/manager/cypress/e2e/core/linodes/create-linode.spec.ts:
```diff
@@ -110,6 +113,9 @@ describe('create linode', () => {
     expect(cy.get('[data-testid="table-row-empty"]').should('not.exist'));
+
+    // Confirm that the GDPR notice is displayed when selecting a region in the EU.
+    cy.findByTestId('eu-agreement-checkbox').should('be.visible');
```
Per offline discussion in the Refinement meeting, EU Consent boxes will only appear until the user has agreed once (i.e., `eu_model: true`), so we'll need to tweak this test to reflect that

jdamore-linode replied:
The E2E tests also operate under the assumption that the EU prompt will never appear, because agreeing to it is one of the steps we take when creating test accounts for CI.

I think a better approach would be a separate test (maybe even in its own spec) which mocks the agreements endpoint and confirms that the prompt appears when European regions are selected, and optionally that it doesn't appear for other regions, and that it doesn't appear when the user has already accepted the agreement.

We can spin that off into a separate task since it's not exactly a small ask, but if you'd rather handle it here I would be happy to help out or even take a stab at it myself cc @jaalah-akamai

jaalah-akamai replied:
Ahh good to know 💡

jaalah-akamai replied:
@jdamore-linode I created a separate ticket for the e2e test. Let's chat offline!

dwiley-akamai reviewed (commented):
I don't think we want to invert the `eu_model` logic. If a user has not previously agreed to the EU Standard Contractual Clauses (i.e., `eu_model: false`), they should see the consent box, and they should not see it if they already agreed to it (i.e., `eu_model: true`).

Right now I never see it, although the account I was on never agreed to the EU Standard Contractual Clauses:

https://github.com/linode/manager/assets/114682940/9b6b9406-e8c4-4c72-be1d-0b7dce184085

abailly-akamai commented on packages/manager/src/components/RegionSelect/RegionSelect.tsx:
```diff
@@ -55,6 +55,9 @@ export const RegionSelect = React.memo((props: RegionSelectProps) => {
     if (selectedId) {
       setSelectedRegion(regionFromSelectedId);
+    } else {
+      // We need to reset the state when create types change
```
@dwiley-akamai no, because the regionSelect refactor isn't in production yet

tyler-akamai commented:
Verified
- [ ] When selected any region selector, ensure that the GDPR checkbox appears for European countries
- [ ] `/linodes/create` (and all the tabs for different create types)
    - [ ] Region selector should clear it's selected state when changing pages
- [ ] When migrating a linode
- [ ] `/volumes/create`
- [ ] `/nodebalancers/create`
- [ ] `/images/create/upload`
- [ ] `/kubernetes/create`
- [ ] `/object-storage/buckets/create`

- [ ] General code review

jaalah-akamai commented:
@dwiley-akamai I realized our akamai accounts default `eu_model` to true which led to the confusion. I reverted the logic, but you'll need to test on a personal account 👍

bnussman-akamai reviewed (commented):
I'm seeing the checkbox on all of the expected create flows. 

The NodeBalancer create flow might need to wrap on small viewports. 

![Screenshot 2023-11-20 at 10 24 48 AM](https://github.com/linode/manager/assets/115251059/75c28aec-1444-4904-bb70-a347288db006)

tyler-akamai commented:
> I'm seeing the checkbox on all of the expected create flows.
> 
> The NodeBalancer create flow might need to wrap on small viewports.
> 
> ![Screenshot 2023-11-20 at 10 24 48 AM](https://user-images.githubusercontent.com/115251059/284314303-75c28aec-1444-4904-bb70-a347288db006.jpg)

Yeah I was curious why on the Linode Create flow, the contract agreement was on its own line and on the NodeBalancer Create flow it was on the same line as the Create button

tyler-akamai commented:
In general the functionality and code LGTM, approving given a few small styling changes.

coliu-akamai reviewed (commented):
Confirmed functionality looks good! Agreed with Banks and Tyler about some styling fixes, had one other question

- ✅  When selected any region selector, ensure that the GDPR checkbox appears for European countries
- ✅ `/linodes/create` (and all the tabs for different create types)
- ✅  Region selector should clear it's selected state when changing pages
- ✅  When migrating a linode
- ✅  `/volumes/create`
- ✅  `/nodebalancers/create`
- ✅  `/images/create/upload`
- ✅  `/kubernetes/create`
- ✅  `/object-storage/buckets/create`
---
Had a quick question - this might be more for the legal teams, but when choosing London, UK as the region, maybe we should change the text to mention whatever the UK laws are instead of the 'EU Standard Contractual Clauses' since UK isn't part of EU anymore: 

![image](https://github.com/linode/manager/assets/139280159/d131bc9f-61c0-4daf-9809-46456bdd6652)

jaalah-akamai commented:
@coliu-akamai Good question, I'll bring that question up to legal folks 👍

jaalah-akamai commented:
@bnussman-akamai @tyler-akamai Seems like it's on it's own line most of the time, so I updated the Nodebalancer structure. 

- Cleaned up styling in Kubernetes checkout, nodebalancer create, and linode create
- Updated `EUAgreementCheckbox` to named export

coliu-akamai reviewed (approved):
Approving - legal question answered in standup 🎉
